### ISSUE TITLE
Support private class properties

### ISSUE THREAD

**Issue by nicholasc** [prettier/prettier#5288]
**Prettier 1.14.3**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAAgMwDYQDuAOlKWHgIYDO1GAKnNTBsKRhgMQAOATgJYA3SvABi-OHgAmSDFACuAWwBGcXgG5S7DJCjNe8sDAi8AFAEpW2mAAt+1AHQ8BwsROkYAvBgCMmqBwAvqTBUCAANCAQ3DD80NTIoJS8vEQACskICSiUghD8UhEgyryUYADWcDAAytxl-FAA5sgwBnCRDdRqMGmljYqUyDiUeF2RAFbUAB4AQqUVVdWUinAAMg1wQyNjIHW8XbzIxZTKAJ4EYZF8DTAA6gW2yAAcAAxXqV23pdxHfExqgk2kV4cAAjvJ+CDepR+oMkMNRu0QF1FPwWm1ItQGo08HAAIryCDwLaIyIwE73KSPJAAJjJpX4eGxAGEIIoBkcoNAgSB5F16CdsgiuoFAkA)
```sh
--parser babylon
```

**Input:**
```jsx
// @flow

class Test {
  #privateField: number;

  constructor() {
  this.#privateField = 1;
  }
}

```

**Output:**
```jsx
SyntaxError: Unexpected token, expected ";" (4:16)
  2 | 
  3 | class Test {
> 4 |   #privateField: number;
    |                ^
  5 | 
  6 |   constructor() {
  7 |   this.#privateField = 1;
```

**Expected behavior:**
It should format the code without any error.

**Comment by lydell** [comment]
Interestingly, when using the Flow parser instead of Babylon the code parses and prints but generates invalid code:

**Prettier 1.14.3**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAAgMwDYQDuAOlKWHgIYDO1GAKnNTBsKRhgMQAOATgJYA3SvABi-OHgAmSDFACuAWwBGcXgG5S7DJCjNe8sDAi8AFAEpW2mAAt+1AHQ8BwsROkYAvBgCMmqBwAvqTBUCAANCAQ3DD80NTIoJS8vEQACskICSiUghD8UhEgyryUYADWcDAAytxl-FAA5sgwBnCRDdRqMGmljYqUyDiUeF2RAFbUAB4AQqUVVdWUinAAMg1wQyNjIHW8XbzIIPhERXwNMADqBbbIABwADJF8EF2XpdxHfExqgpuRvDgAEd5PxAb1KP1BkhhqN2iAuop+C02pFqA1Gng4ABFeQQeBbOGRGCUZTXKS3JAAJmJpX4eAxAGEIIoBkcoNB-iB5F16KTsrCuoFAkA)
```sh
--parser flow
```

**Input:**
```jsx
// @flow

class Test {
  #privateField: number;

  constructor() {
  this.#privateField = 1;
  }
}

```

**Output:**
```jsx
// @flow

class Test {
  privateField: number;

  constructor() {
    this.#privateField = 1;
  }
}

```

**Second Output:**
```jsx
SyntaxError: Private fields must be declared before they can be referenced. `#privateField` has not been declared. (7:10)
   5 | 
   6 |   constructor() {
>  7 |     this.#privateField = 1;
     |          ^^^^^^^^^^^^^
   8 |   }
   9 | }
  10 | 
```

**Comment by j-f1** [comment]
The first issue seems to be a problem with Babylon, and the second is a bug in our printer.

**Comment by duailibe** [comment]
The second is because Flow AST is `ClassProperty { key: { Identifier "privateField" } }` whereas the Babylon should be `ClassProperty { key: { PrivateName { Identifier "privateField" } } }` ("pseudo-syntax")

**Comment by fisker** [comment]
This is fixed by #7484 , also tests covered

**Prettier pr-7484**
[Playground link](https://deploy-preview-7484--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEB6VACAAgMwDYQDuAOlKWHgIYDO1GAKnNTBsKRhgMQAOATgJYA3SvABi-OHgAmSDFACuAWwBGcXgG5S7DJCjNe8sDAi8AFAEpW2mAAt+1AHQ8BwsROkYAvBgCMmqBwAvqSBIAA0IBDcMPzQ1MiglLy8RAAKSQjxKJSCEPxS4SDKvJRgANZwMADK3KX8UADmyDAGcBE2MIp4AOp28NS1YHBVmfwxgmMAnsjgtIX11GowqSUNipTIOJR4ixEAVtQAHgBCJeWVVZSKcAAy9XCb27sgB4dV9Q14cACK8hDwjx2bRAtV4i14M3wREKfHqMG6+VsyAAHAAGCJ8CCLbolbgzPhMNSCB4RACOf3gKyiWRANAAtFA4HApMzCrw4OT+OyVpQ1hskFsgRFFop+M1WsKPl9fv8HgKnsCYJRlAipEikAAmCItSj8PAfADCEEU6xmUGgJJA8kW9GVWUFz0E8jgAEkoCzYFUwAJogBBd1VGCTL6AxaBQJAA)
```sh
--arrow-parens avoid
--parser flow
--trailing-comma none
```

**Input:**
```jsx
// @flow

class Test {
  #privateField: number;

  constructor() {
  this.#privateField = 1;
  }
}
```

**Output:**
```jsx
// @flow

class Test {
  #privateField: number;

  constructor() {
    this.#privateField = 1;
  }
}

```

**Comment by evilebottnawi** [comment]
Fixed https://github.com/prettier/prettier/pull/7484
